docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,253 @@
+system would be unhelpfully abstract at this stage.
+
+## Decisions
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects.
+
+  - Reactors should not need to expose their interfaces *within* the
+	implementation of the node type, except in the context of some groups of
+	services.
+
+  - This refactoring should be entirely opaque to users.
+
+  - These node initialization changes should not require a
+	reevaluation of the `service.Service` or a generic initialization
+	orchestration framework.
+
+- If required for the 0.35 release, add an "unsafe" way to construct a
+  node service with user-supplied mempool, including making the
+  relevant interfaces and types external/exported.
+
+- Prioritize implementation of p2p-based statesync reactor to obviate
+  need for users to inject a custom state-sync provider.
+
+## Detailed Design
+
+The [current
+nodeImpl](https://github.com/tendermint/tendermint/blob/master/node/node.go#L47)
+includes direct references to the implementations of each of the
+reactors, which should be replaced by references to `service.Service`
+objects. This will require moving the construction of the [rpc
+service](https://github.com/tendermint/tendermint/blob/master/node/node.go#L771)
+into the constructor of
+[makeNode](https://github.com/tendermint/tendermint/blob/master/node/node.go#L126).
+
+In order to prevent adding complexity to the `node` package, this
+project will add an implementation to the service `service` package
+that implements `service.Service` and is comprised of a sequence of
+underlying `service.Service` objects and handles their
+startup/shutdown in a consistent order.
+
+Consensus, blocksync (nee fast sync.), and statesync all depend on
+each other, and have significant initialization dependencies that are
+presently encoded in the `node` package, and as part of this change,
+a new package/component will encapsulate the initialization of this
+area of functionality.
+
+In order to support replacement of a component, a new public function
+will be added to the public interface of `node` with a signature that
+resembles the following:
+
+```go
+func NewWithServices(conf *config.Config,
+	logger log.Logger,
+	cf proxy.ClientCreator,
+	gen *types.GenesisDoc,
+	srvs ...service.Service
+) (service.Service, error) {
+```
+
+The `service.Service` objects will be initialized in the order their
+supplied, after all pre-configured/default services have started (and
+shut down first, comparatively.) If any of the services implement
+additional interfaces, that allow them to replace specific default
+services. `NewWithServices` will validate input service lists with the
+following rules:
+
+- no running services
+- all values of `service.Service.String()` must be unique.
+- for replacing default services, callers cannot supply more than
```

Comment:
Can you give an example of what this may mean? I think my preference is something more linkenable to what `net.http` does in regards to its `Transport` type. It provides a default transport type:
https://github.com/golang/go/blob/master/src/net/http/transport.go#L43

And when no tansport is set on an object, it uses this default: https://github.com/golang/go/blob/aa4e0f528e1e018e2847decb549cfc5ac07ecf20/src/net/http/client.go#L58-L62

I find this to be a bit more sane than using a system of replacing a component by key or by using reflect. Using reflect or keys to replace components likely mean your API will now require the user to understand the code inside of your constructor and they will also need to right a test to make sure that their type is always correctly loaded in in case they change something about it.

Reply:
I think we can drop this requirement: this does mean that we'll end up needing to export interfaces *other* than `service.Service` for reactors (which might be required anyway?) which means we'll just have a type switch rather than a switch on the service name, and that's fine.